buildscript.py

```diff
@@ -0,0 +1,34 @@
+import os
+import redash_stmo
+from subprocess import call
+from distutils.dir_util import copy_tree
+
+from pkg_resources import iter_entry_points, resource_filename
+
+
+# Make a directory for extensions and set it as an environment variable
+# to be picked up by webpack.
+EXTENSIONS_RELATIVE_PATH = "client/app/extensions"
+EXTENSIONS_DIRECTORY = os.path.join(os.getcwd(), EXTENSIONS_RELATIVE_PATH)
+if not os.path.exists(EXTENSIONS_DIRECTORY):
+    os.makedirs(EXTENSIONS_DIRECTORY)
+os.environ["EXTENSIONS_DIRECTORY"] = EXTENSIONS_RELATIVE_PATH
+
+for entry_point in iter_entry_points('webpack.bundles'):
+	extension_data = entry_point.load()
+
+	# This is where the frontend code for an extension lives
+	# inside of its package.
+	content_folder_relative = os.path.join(
+		extension_data['extension_directory'],
+		extension_data['frontend_content'])
+	content_folder = resource_filename(redash_stmo.__name__, content_folder_relative)
+
+	# This is where we place our extensions folder.
+	destination = os.path.join(
+		EXTENSIONS_DIRECTORY,
+		extension_data['extension_directory'])
+
+	copy_tree(content_folder, destination)
+
+call("npm run build", shell=True)
```

Comment:
Let's remove the call to `npm run build` here and instead add a new item to the `scripts` section in the `package.json` file called `prebuild` with the content `bin/bundle-extensions` so it's automatically called before the `build` script in `package.json`.

For the `watch` script in package.json we could also add a `prewatch` step that calls `bin/bundle-extensions` with a `--link` option which would create symlinks instead of copies. I think. What do you think?

Reply:
There are a few problems I came across here with this:

1. `npm run build` is called from `docker-compose build` when building the `Dockerfile` and this adds a new `/app/client/app/extensions` directory in the container. However the `docker-compose.yml` file then overrides this by mounting the volume `.:/app` since locally the `extensions` folder is not yet created. This results in the `extensions` folder not being visible as expected when running `docker-compose build`.

2. `npm run build` won't be callable locally because the `bundle-extensions` requires `redash-stmo` to be installed so that it's contents are copied over.

For these two reasons I made this buildscript a separate executable (https://github.com/mozilla/redash/pull/463/files#diff-76ffdf10e3a9d8396a3951f272eaf403R75) that did not run from the `Dockerfile` or locally. 

Is there perhaps some Docker magic that you know of to get around this?

In terms `prewatch`, I'm not quite sure what the purpose of this would be. Could you elaborate a bit? It seemed that once the `client/app/extensions` directory exists, changes in it it will already be capturd by `watch`.

Reply:
I think this might work if the docker-compose.yml `command` field is used to run this buildscript as well as the Dockerfile. I will try that

Reply:
Okay, to summarize our discussion just now. The problem that you hinted at in 1. and 2. is basically the result of an inconsistencies of build environment between when the image is build (and calling npm run build) and when a developer is updating the build from outside the container based on that image.

On top of that docker-compose overlaps the folder that is created inside the image when mounting the working directory over it. As you correctly said, that's why Redash asks developers to run `npm build` from the host machine, since that's the only placen that has access to the build directory during development.

A strategy to solve that inconsistency between image and development (or: disparity between development and production environment) is:

1. move the folders that are used by both outside the path where docker-compose is mounting the working directory

   For ATMO we solved this by [creating a new separate directory and installing npm modules there](https://github.com/mozilla/telemetry-analysis-service/blob/450587d8df0b1b68688c1b6aadca138dc2b4cb68/Dockerfile#L39-L56). And then run the build process both when the [image is created](https://github.com/mozilla/telemetry-analysis-service/blob/450587d8df0b1b68688c1b6aadca138dc2b4cb68/Dockerfile#L63-L64) as well as [during development](https://github.com/mozilla/telemetry-analysis-service/blob/master/bin/run#L40-L41).

2. run all build processes (npm build and bundle-extensions) inside the container for both image creation and development

3. provide helpers such as a script that can be called from the host machine to trigger the container-side build process to help counteract the impact on the development workflow

So I think we only need `/opt/npm` for the NPM modules and then provide a docker entrypoint command that provides the ability to run bundle-extensions from the host machine via docker-compose, e.g. what the user would call to build the bundle including the extensions: `docker-compose run server bundle`.